File: regression/esbmc-cpp/projects/mailutils-2.2/examples/cpp/lsf.cc
/* GNU Mailutils -- a suite of utilities for electronic mail
   Copyright (C) 2009, 2010 Free Software Foundation, Inc.

   GNU Mailutils is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   GNU Mailutils is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with GNU Mailutils; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
   MA 02110-1301 USA */

#include <iostream>
#include <cstdlib>
#include <mailutils/cpp/mailutils.h>

using namespace std;
using namespace mailutils;

static int
enumfun (mu_folder_t folder, struct mu_list_response *resp, void *data)
{
  printf ("%c%c %c %4d %s\n",
	  (resp->type & MU_FOLDER_ATTRIBUTE_DIRECTORY) ? 'd' : '-',
	  (resp->type & MU_FOLDER_ATTRIBUTE_FILE) ? 'f' : '-',
	  resp->separator,
	  resp->level,
	  resp->name);
  return 0;
}

int
ls_folders (const string& fname, const string& ref, void* pattern, int level)
{
  try {
    Folder folder (fname);
    folder.open ();

    List list = folder.enumerate (ref, pattern, 0, level, enumfun, NULL);
    cout << "Number of folders: " << list.count () << endl;

    folder.close ();
  }
  catch (Exception& e)
  {
    if (e.status () == MU_ERR_NOENT) {
      cout << "No folders matching " << ref << " " << pattern
	   << " in " << fname << endl;
    }
    else {
      cerr << e.method () << ": " << e.what () << endl;
      return 1;
    }
  }
  return 0;
}

int
main (int argc, char *argv[])
{
  string folder;
  string ref;
  const char *pattern = "*";
  int level = 0;

  switch (argc)
    {
    case 5:
      level = atoi (argv[4]);
    case 4:
      pattern = argv[3];
    case 3:
      ref = argv[2];
    case 2:
      folder = argv[1];
      break;
    default:
      error ("usage: lsf folder [ref] [pattern] [recursion-level]\n");
      return 1;
    }

  register_all_mbox_formats ();

  return ls_folders (folder, ref, (char *)pattern, level);
}

File: regression/esbmc-cpp/projects/mailutils-2.2/examples/python/lsf.py
#  GNU Mailutils -- a suite of utilities for electronic mail
#  Copyright (C) 2009, 2010 Free Software Foundation, Inc.
#
#  GNU Mailutils is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3, or (at your option)
#  any later version.
#
#  GNU Mailutils is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with GNU Mailutils; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301 USA

import sys
from mailutils import folder
from mailutils import registrar
from mailutils.error import *

def ls_folders (fname, ref, pattern, level):
    try:
        fld = folder.Folder (fname)
        fld.open ()

        list = fld.list (ref, pattern, level)
        for f in list:
            print f
        print "Number of folders: %d" % len (list)

        fld.close ()
    except Error, e:
        print e

if __name__ == '__main__':
    pattern = "*"
    level = 0
    argc = len (sys.argv)

    if argc == 5:
        level = int (sys.argv[4])
        pattern = sys.argv[3]
        ref = sys.argv[2]
        fname = sys.argv[1]
    elif argc == 4:
        pattern = sys.argv[3]
        ref = sys.argv[2]
        fname = sys.argv[1]
    elif argc == 3:
        ref = sys.argv[2]
        fname = sys.argv[1]
    elif argc == 2:
        ref = None
        fname = sys.argv[1]
    else:
        print "usage: lsf folder [ref] [pattern] [recursion-level]"
        sys.exit (0)

    registrar.register_format ()
    ls_folders (fname, ref, pattern, level)
